### PR TITLE
fix(common): infer correct type when `trackBy` is used in `ngFor`

### LINKED ISSUE #40125: trackBy incorrectly changes item type
# 🐞 bug report

### Affected Package

The issue is caused by package @angular/core

### Description

Using the ngFor trackBy function sets the item type to exactly the type of its argument

In the example below, use of the trackByBase function widens (i.e makes less specific) `item` in the template up to `Base`, even though it actually is a `Derived`. This is unexpected; the ngFor source array is already typed and the expectation is that the inferred type in the template is the type of the array, not the type of the trackBy argument.

## 🔬 Minimal Reproduction

Stackblitz: https://stackblitz.com/edit/angular-ivy-urqcqm

~~~
import { Component, VERSION } from '@angular/core';


interface Base {
    id: string;
}


interface Derived extends Base {
    name: string;
}


const DERIVED_ARRAY: Derived[] = [
    { id: 'one', name: 'First' },
    { id: 'two', name: 'Second' },
];


@Component({
    selector: 'my-app',
    templateUrl: './app.component.html',
    styleUrls: [ './app.component.css' ]
})
export class AppComponent  {
    array = DERIVED_ARRAY;

    trackByBase(index: number, item: Base): string {
        return item.id;
    }
}
~~~
and template
~~~
<div *ngFor="let item of array; trackBy:trackByBase">
    {{ item.name }}
</div>
~~~

## 🔥 Exception or Error

<pre><code>
Error in src/app/app.component.html (10:13)
Property 'name' does not exist on type 'Base'.
</code></pre>

## 🌍  Your Environment

**Angular Version:**

<pre><code>
     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 11.0.3
Node: 12.18.4
OS: linux x64

Angular: 11.0.3
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, localize, platform-browser
... platform-browser-dynamic, router, upgrade
Ivy Workspace: Yes

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1100.3 (cli-only)
@angular-devkit/build-angular   0.1100.3
@angular-devkit/core            11.0.3
@angular-devkit/schematics      11.0.3
@angular/cdk                    11.0.2
@angular/material               11.0.2
@schematics/angular             11.0.3
@schematics/update              0.1100.3
ng-packagr                      11.0.3
rxjs                            6.6.3
typescript                      4.0.5
</code></pre>

### PR DESCRIPTION
**Pull request by JoostK**
See individual commits

Fixes #40125

### CODE REVIEW

**Review by jessicajaniuk — APPROVED** [review]
LGTM 🍪 

reviewed-for: public-api

File: packages/core/src/change_detection/differs/iterable_differs.ts [hunk 1]
```diff
@@ -121,7 +121,7 @@ export interface IterableChangeRecord<V> {
  * @publicApi
  */
 export interface TrackByFunction<T> {
-  (index: number, item: T): any;
+  <U extends T>(index: number, item: U): any;
```
Comment (jelbourn) [inline packages/core/src/change_detection/differs/iterable_differs.ts:124]: Could you add a comment here that adds context to  `<U extends T>` with a link to the issue?

**Review by jelbourn — APPROVED** [review]
LGTM

Reviewed-for: public-api

**Review by petebacondarwin — COMMENTED** [review]
Reviewed-for: public-api

**Review by petebacondarwin — APPROVED** [review]
Reviewed-for: public-api

**Review by jessicajaniuk — APPROVED** [review]
LGTM :cookie:

reviewed-for: public-api

**Review by alxhub — APPROVED** [review]
Reviewed-for: public-api

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:374115254:BASE:374115261:1621223097711:253baff9).

**Comment by AndrewKushnir** [comment]
FYI, I'm adding the "blocked" label for now, since there are few broken targets in Google's codebase identified by the presubmit. More investigation is needed to see if apps code or the code of this PR should be updated. I will perform the necessary investigation and post an update (will try to create a minimal repro if needed). Thank you.

**Comment by AndrewKushnir** [comment]
Also started a [global presubmit](https://test.corp.google.com/OCL:374115254:BASE:374328782:1621304365700:6f509) to see if there are other apps with similar usage patterns.

**Comment by JoostK** [comment]
@AndrewKushnir This one is ready for another round of presubmit.

**Comment by AndrewKushnir** [comment]
[Global Presubmit #2](https://test.corp.google.com/ui#id=OCL:377198431:BASE:377198478:1622687059408:24ed539f).

**Review by AndrewKushnir — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
@JoostK presubmits (including a global one) went well and this PR looks ready for merge (just needs the "merge" label).

**Pull request closed without merging**